File: c++Poker/Deck.cpp
#include <iostream>
#include "Card.h"
#include <list>
#include <algorithm>
#include <random>
#include <ctime>
using namespace std;

class Deck
{
    private:
        list <Card> deck;
    public:
        
        Deck()
        {  
        
            // Initialize the deck with cards
            for (int suit = 1; suit <= 4; suit++)
            {
                for (int face = 1; face <= 13; face++)
                {
                    Card card(face, suit);
                    deck.push_back(card);
                }
            }
        }

        void shuffle()
        {
            srand(static_cast<unsigned>(time(nullptr)));
            vector<Card> cardVector(deck.begin(), deck.end());

            for (int i = cardVector.size() - 1; i > 0; i--)
            {
                int j = rand() % (i + 1);
                swap(cardVector[i], cardVector[j]);
            }

            deck.assign(cardVector.begin(), cardVector.end());
        }

        Card popCard()
        {
            if (!deck.empty())
            {
                Card topCard = deck.front();
                deck.pop_front();
                return topCard;
            }
            else
            {
                // Handle the case where the deck is empty
                // You can throw an exception or return a special "no card" indicator.
                // For simplicity, we'll just return an empty card here.
                cout<<"Deck is empty";
                return Card(0, 0);
            }
        }

        void showDeck()
        {
            int i=0;
            for(Card c: deck)
            {
                c.printCard();
                cout<<" ";
                i++;
                if(i==13)
                {
                    cout<<endl;
                    i=0;
                }
            }
            cout<<endl;
        }

};

File: goPoker/deck.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)


type Deck struct {
	Cards []Card
}

func NewDeck() Deck {
	deck := Deck{}
	// Initialize the deck with cards
	for suit := 1; suit <= 4; suit++ {
		for face := 1; face <= 13; face++ {
			card := Card{Face: face, Suit: suit}
			deck.Cards = append(deck.Cards, card)
		}
	}
	return deck
}

func (d *Deck) Shuffle() {
	rand.Seed(time.Now().UnixNano())
	rand.Shuffle(len(d.Cards), func(i, j int) {
		d.Cards[i], d.Cards[j] = d.Cards[j], d.Cards[i]
	})
}

func (d *Deck) PopCard() Card {
	if len(d.Cards) > 0 {
		topCard := d.Cards[0]
		d.Cards = d.Cards[1:]
		return topCard
	} else {
		// Handle the case where the deck is empty
		// For simplicity, returning an empty card here.
		fmt.Println("Deck is empty")
		return Card{}
	}
}

func (d Deck) ShowDeck() {
	for i, c := range d.Cards {
		c.printCard()
		fmt.Print(" ")
		if (i+1)%13 == 0 {
			fmt.Println()
		}
	}
	fmt.Println()
}